specs/_features/eip7732/beacon-chain.md

```diff
@@ -0,0 +1,716 @@
+    latest_execution_payload_header: ExecutionPayloadHeader
+    # Withdrawals
+    next_withdrawal_index: WithdrawalIndex
+    next_withdrawal_validator_index: ValidatorIndex
+    # Deep history valid from Capella onwards
+    historical_summaries: List[HistoricalSummary, HISTORICAL_ROOTS_LIMIT]
+    deposit_requests_start_index: uint64
+    deposit_balance_to_consume: Gwei
+    exit_balance_to_consume: Gwei
+    earliest_exit_epoch: Epoch
+    consolidation_balance_to_consume: Gwei
+    earliest_consolidation_epoch: Epoch
+    pending_balance_deposits: List[PendingBalanceDeposit, PENDING_BALANCE_DEPOSITS_LIMIT]
+    pending_partial_withdrawals: List[PendingPartialWithdrawal, PENDING_PARTIAL_WITHDRAWALS_LIMIT]
+    pending_consolidations: List[PendingConsolidation, PENDING_CONSOLIDATIONS_LIMIT]
+    # PBS
+    latest_block_hash: Hash32  # [New in EIP-7732]
+    latest_full_slot: Slot  # [New in EIP-7732]
+    latest_withdrawals_root: Root  # [New in EIP-7732]
+```
+
+## Helper functions
+
+### Math
+
+#### `bit_floor`
+
+```python
+def bit_floor(n: uint64) -> uint64:
+    """
+    if ``n`` is not zero, returns the largest power of `2` that is not greater than `n`.
+    """
+    if n == 0:
+        return 0
+    return uint64(1) << (n.bit_length() - 1)
+```
+
+### Misc
+
+#### `remove_flag`
+
+```python
+def remove_flag(flags: ParticipationFlags, flag_index: int) -> ParticipationFlags:
+    flag = ParticipationFlags(2**flag_index)
+    return flags & ~flag
+```
+
+### Predicates
+
+#### `is_valid_indexed_payload_attestation`
+
+```python
+def is_valid_indexed_payload_attestation(
+        state: BeaconState, 
+        indexed_payload_attestation: IndexedPayloadAttestation) -> bool:
+    """
+    Check if ``indexed_payload_attestation`` is not empty, has sorted and unique indices and has
+    a valid aggregate signature.
+    """
+    # Verify the data is valid
+    if indexed_payload_attestation.data.payload_status >= PAYLOAD_INVALID_STATUS:
+        return False
+
+    # Verify indices are sorted and unique
+    indices = indexed_payload_attestation.attesting_indices
+    if len(indices) == 0 or not indices == sorted(set(indices)):
+        return False
+
+    # Verify aggregate signature
+    pubkeys = [state.validators[i].pubkey for i in indices]
+    domain = get_domain(state, DOMAIN_PTC_ATTESTER, None)
+    signing_root = compute_signing_root(indexed_payload_attestation.data, domain)
+    return bls.FastAggregateVerify(pubkeys, signing_root, indexed_payload_attestation.signature)
+```
+
+#### `is_parent_block_full`
+
+This function returns true if the last committed payload header was fulfilled with a payload, this can only happen when both beacon block and payload were present. This function must be called on a beacon state before processing the execution payload header in the block. 
+
+```python
+def is_parent_block_full(state: BeaconState) -> bool:
+    return state.latest_execution_payload_header.block_hash == state.latest_block_hash
```

Comment:
In tests, `block_hash` may be zero hash, and then this has false positives.

In attestations the check simply compares `slot.latest_full_slot`, so I guess one could just replace this with `state.latest_full_slot == state.latest_block_header.slot`.

Reply:
good point!